File: goodArray.java
// Asked in Goldman Sachs interview
// https://www.desiqna.in/6085/goldman-coding-question-solution-engineering-campus-program

// https://leetcode.com/problems/range-product-queries-of-powers/discuss/2740795
// /simple-java-solution
import java.util.*;
class Solution{ // TC O(n*m) SC O(n)
	public List<Integer> goodArray(int n, List<List<Integer>> queries){
		List<Integer> equivalentBinary= new ArrayList<>();
		List<Integer> goodArray= new ArrayList<>();
		while(n>0){
			equivalentBinary.add(n%2);
			n=n/2;
		}

		for(int i=0; i<equivalentBinary.size();i++){
			if(equivalentBinary.get(i)==1){
				goodArray.add(1<<i); 
				//goodArray.add((int)Math.pow(2,i));
			}
		}

		System.out.println(goodArray);
		
		List<Integer> answer= new ArrayList<>();
		for(List<Integer> list: queries){
			int l=list.get(0);
			int r=list.get(1);
			int m=list.get(2);
			int through=1;
			while (l<=r){
				through*=goodArray.get(l-1);
				l++;
			}
			answer.add(through%m);
		}

		return answer;
	}

}


class goodArray{
    public static void main(String[] args){
      List<List<Integer>> queries= new ArrayList<>(Arrays.asList(Arrays.asList(1,2,4),
      	Arrays.asList(2,2,8),Arrays.asList(1,1,4)));
      int n=6;
      Solution s= new Solution();
      System.out.println(s.goodArray(n,queries));
    }
    
}

File: goodArray.py
# Asked in Goldman Sachs interview
# https://www.desiqna.in/6085/goldman-coding-question-solution-engineering-campus-program
def goodArray(n,queries): #TC O(n*m) // SC O(n)
	equivalentBinary=[]
	goodArray=[]
	while(n>0):
		equivalentBinary.append(n%2);
		n=n//2

	for i in range(len(equivalentBinary)):
		if(equivalentBinary[i]==1):
			goodArray.append(1<<i)
			#goodArray.append(pow(2,i))

	print(goodArray)
	answer=[]
	for query in queries:
		l=query[0];
		r=query[1];
		m=query[2];
		through=1
		while l<=r:
			through*=goodArray[l-1]
			l+=1

		answer.append(through%m)

	return answer 


print(goodArray(6,[[1,2,4],[2,2,8],[1,1,4]]))
print(goodArray(26,[[1,2,1009],[3,3,5]]))
print(goodArray(12,[[1,2,84],[2,2,3]]))